File: v2/63.py





class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        # 6 star, 动态规划， 注意要特殊处理最上和左的行和列
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        dp = [[0 for _ in range(n)] for _ in range(m)]
        for i in range(m):
            if obstacleGrid[i][0] == 0:
                dp[i][0] = 1
            else:
                break
        for j in range(n):
            if obstacleGrid[0][j] == 0:
                dp[0][j] = 1
            else:
                break

        for i in range(1, n):
            for j in range(1, m):
                if obstacleGrid[j][i]:
                    dp[j][i] = 0
                else:
                    dp[j][i] = dp[j-1][i] + dp[j][i-1]
        return dp[-1][-1]

# a = [[0,0,0],
#      [1,0,0],
#      [0,0,0]]
# # a = [[1]]
# print(Solution().uniquePathsWithObstacles(a))

File: go/63.go

//package main

//import "fmt"

// 6 star, 动态规划
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	if obstacleGrid[0][0] == 1 {return 0}
	cols := len(obstacleGrid)
	rows := len(obstacleGrid[0])
	dp := make([][]int, cols)
	for i:=0; i<cols; i++ {
		dp[i] = make([]int, rows)
		if i == 0 {dp[i][0] = 1}
		if (i > 0 && dp[i-1][0] == 0) || obstacleGrid[i][0] == 1 {
			dp[i][0] = 0
		} else {
			dp[i][0] = 1
		}

	}
	for j:=0; j<rows; j++ {
		if (j > 0 && dp[0][j-1] == 0) || obstacleGrid[0][j] == 1 {
			dp[0][j] = 0
		} else {
			dp[0][j] = 1
		}
	}
	for i:=1;i<cols;i++ {
		for j:=1; j<rows; j++ {
			if obstacleGrid[i][j] == 1{
				dp[i][j] = 0
			} else {
				dp[i][j] = dp[i-1][j] + dp[i][j-1]
			}

		}
	}
	return dp[cols-1][rows-1]

}


//func main() {
//	nums := make([][]int, 3)
//	nums[0] = []int {0,0,0}
//	nums[1] = []int {0,1,0}
//	nums[2] = []int{0,0,0}
//	fmt.Println(uniquePathsWithObstacles(nums))
//
//}
